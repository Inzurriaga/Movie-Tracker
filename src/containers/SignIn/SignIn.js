import React, { Component } from 'react';
import { connect } from 'react-redux';
import { user, initialFavorites } from '../../actions';
import { postFetch, getFetch } from '../../api';
import PropTypes from 'prop-types';


export class SignIn extends Component {
  constructor(props){
    super(props)
    this.state = {
      name: '',
      email: '',
      password: '',
      newUser: true,
      error: '',
    }
  }

  handleLogin = async (url, method, body) => {
    try {
      const response = await postFetch(url, method, body);
      const retrievedData = await response.json()
      if(response.ok){
        console.log("if im working", retrievedData)
        return retrievedData;
      }
    } catch (error) {
      console.log("im not working in the hanlde")
      this.setState({
        error: "not working"
      })
    }
  }

  signIn = async (event) => {
    console.log("sign in")
    event.preventDefault();
    this.setState({error: ''})
    const { password, email } = this.state
    const url = 'users'
    const retrievedData = await this.handleLogin(url, 'POST', { password, email })
    if(this.state.error !== "not working" ) {
      this.props.user(retrievedData.data);
      await this.fetchUserFavorite(this.props.userInfo.id);
      this.props.hideModal()
    }
  }

  fetchUserFavorite = async (id) => {
    console.log("id", id)
    const url = `users/${id}/favorites`
    const response = await getFetch(url)
    this.props.initialFavorites(response)
  }

  createAccount = async (event) => {
    event.preventDefault();
    const { name, password, email } = this.state
    const url = 'users/new'
    this.setState({error: ''})
    await this.handleLogin(url, 'POST', { name, password, email })
    if(this.state.error !== "not working" ) {
      console.log("create")
      this.signIn(event)
    }
  }

  userInput = (event) => {
    const value = event.target.value
    const key = event.target.classList
    this.setState({
      [key]: value
    })
  }

  handleSignSwitch = () => {
    this.setState({
      newUser: !this.state.newUser,
      error: ''
    })
  }


  render(){
    const btnActive = 'btnChoice choice-active';
    const btnInactive = 'btnChoice choice-inactive';
    const { newUser } = this.state;
    return (
      <div className="SignIn-Content">
        <section className="Modal-Close">
          <button className="closeBtn" onClick={this.props.hideModal}>X</button>
        </section>
        <section className="Modal-Header">
          <button className={newUser ? btnInactive : btnActive} onClick={this.handleSignSwitch}>Sign In</button>
          <button className={newUser ? btnActive : btnInactive} onClick={this.handleSignSwitch}>Create Account</button>
        </section>
        <section className="Modal-Body">
        { !this.state.newUser ?
            <form onSubmit={this.createAccount} className="CreateForm">
              <label>Name</label>
              <input onChange={this.userInput} type="text" className="name"/>
              <label>Email</label>
              <input onChange={this.userInput} type="text" className="email"/>
              <label>Password</label>
              <input onChange={this.userInput} type="text" className="password"/>
              { this.state.error === "not working" && 
              <p>Email has already been used</p>
              }
              <button type="submit">Create Account</button>
            </form>
          :
          <form onSubmit={this.signIn} className="SignForm">
            <label>Email</label>
            <input onChange={this.userInput} type="text" className="email"/>
            <label>Password</label>
            <input onChange={this.userInput} type="text" className="password"/>
            { this.state.error === "not working" && 
              <p>Email and Password do not match</p>
            }
            <button type="submit">Sign In</button>
          </form>
        }
        </section>
      </div>
    )
  }
}


SignIn.propTypes = {
  className: PropTypes.string.isRequired,
  hideModal: PropTypes.func.isRequired,
  initialFavorites: PropTypes.func.isRequired,
  show: PropTypes.bool.isRequired,
  user: PropTypes.func.isRequired,
  userInfo: PropTypes.object.isRequired
}



export const mapStateToProps = (state) => ({
  userInfo: state.user,
})

export const mapDispatchToProps = (dispatch) => ({
  user: (userInfo) => dispatch(user(userInfo)),
  initialFavorites: (favorites) => dispatch(initialFavorites(favorites)),
})

export default connect(mapStateToProps, mapDispatchToProps)(SignIn);
