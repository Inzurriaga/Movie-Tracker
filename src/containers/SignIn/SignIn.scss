.SignIn-Content {
  position:fixed;
  background: white;
  height: auto;
  padding: 1em;
  top:50%;
  left:50%;
  transform: translate(-50%,-50%);

  .Modal-Close {
    .closeBtn {
      position: absolute;
      right: -5px;
      top: -5px;
      background: none;
      border:none;
      outline: none;
    }
  }


  .Modal-Header {
    padding-bottom: 1em;
    display: flex;
    width: 100%;
    box-sizing: border-box;

    .btnChoice {
      width: 50%;
      outline: none;

      &.choice-active {
        background:#fff;
      }

      &.choice-inactive {
        background:#cccccc;
        pointer-events: none;
      }

    }
  }


  .Modal-Body {

    form {
      display: flex;
      flex-direction: column;

      label {

      }

      button {

      }

    }
  }

}
