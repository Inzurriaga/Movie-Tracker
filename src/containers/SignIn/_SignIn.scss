.SignIn-Content {
  position:fixed;
  background: #2f2f2f;
  width: 100%;
  max-width: 350px;
  border-radius: 5px;
  height: auto;
  padding: 0 1.5em;
  top: 50%;
  left: 50%;
  transform: translate(-50%,-50%);
  box-shadow: 0px 2px 10px 3px #0000006b;

  .Modal-Close {

    .closeBtn {
      position: absolute;
      right: 0px;
      top: 0px;
      background: none;
      border: none;
      outline: none;
      color: #a7a7a7;
      background: #3e3e3e;
      font-size: 1.1em;
      line-height: 1.5;
      cursor: pointer;
      transition: all .5s ease;
      border-top-right-radius: 5px;

      &:hover {
        color: #d2d2d2;
        background: #4e4e4e;
      }
    }
  }


  .Modal-Header {
    padding-top: 1.5em;
    padding-bottom: 1.5em;
    display: flex;
    width: 100%;
    box-sizing: border-box;
    justify-content: center;

    .btnChoice {
      width: 40%;
      padding: .5em;
      font-size: 1em;
      outline: none;
      text-transform: uppercase;
      font-family: $font-Sans;
      border:none;
      cursor: pointer;
      transition: all .5s ease;

      &.choice-active {
        background: #272635;
        color: #505050;
        box-shadow: 0px 0px 4px 0px #00000070;

        &:hover {
          color: #505050;
        }
      }

      &.choice-inactive {
        background: #231f20;
        color: #505050;
        pointer-events: none;
      }

    }
  }


  .Modal-Body {
    padding-bottom: 1.5em;
    form {
      display: flex;
      flex-direction: column;

      label {
        font-family: $font-Sans;
        color: #7d7d7d;
        font-weight: 100;
        font-size: .9em;
        letter-spacing: .1em;
        padding-bottom: .4em;
        outline: none;
      }

      input {
        margin-bottom: 1em;
        line-height: 1;
      }

      button {
        margin-top: 0.5em;
        font-family: $font-Sans;
        background: #272635;
        color: #dadada;
        text-transform: uppercase;
        box-shadow: 0px 0px 4px 0px #00000070;
        font-weight: 100;
        border: none;
        outline: none;
        padding: .5em 0;
        border-radius: 2px;
        box-shadow: 0px 1px 4px 0px black;
        cursor: pointer;
        transition: all .5s ease;

        &:hover {
          background: #888333;
          color: #dedede;
        }
      }

    }
  }

}
