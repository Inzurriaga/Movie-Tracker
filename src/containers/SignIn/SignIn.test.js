import React from 'react';
import { SignIn, mapStateToProps } from './SignIn';
import { shallow } from 'enzyme';


describe('Sign In', () => {
  let signInComponent;

  describe('SignIn Component', () => {

    beforeEach(() => {
      signInComponent = shallow(<SignIn />)
    })

    it.skip('should match the snapshot', () => {

      expect(signInComponent).toMatchSnapshot()

    })

    it.skip('should have default properties', () => {
      expect(signInComponent.state()).toEqual({
        name: '',
        email: '',
        password: '',
        newUser: true,
      });

    })

  })


  describe('SignIn mapStateToProps', () => {

    it('should return an object with genres information', () => {

    })

  })

})
