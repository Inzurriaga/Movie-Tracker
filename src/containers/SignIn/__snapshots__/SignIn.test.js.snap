// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign In SignIn Component should match the snapshot 1`] = `
<div
  className="SignIn-Content"
>
  <section
    className="Modal-Close"
  >
    <button
      className="closeBtn"
      onClick={[MockFunction]}
    >
      X
    </button>
  </section>
  <section
    className="Modal-Header"
  >
    <button
      className="btnChoice choice-inactive"
      onClick={[Function]}
    >
      Sign In
    </button>
    <button
      className="btnChoice choice-active"
      onClick={[Function]}
    >
      Create Account
    </button>
  </section>
  <section
    className="Modal-Body"
  >
    <form
      className="SignForm"
      onSubmit={[Function]}
    >
      <label>
        Email
      </label>
      <input
        className="email"
        onChange={[Function]}
        type="text"
      />
      <label>
        Password
      </label>
      <input
        className="password"
        onChange={[Function]}
        type="text"
      />
      <button
        type="submit"
      >
        Sign In
      </button>
    </form>
  </section>
</div>
`;

exports[`Sign In SignIn Component should match the snapshot when newUser becomes false 1`] = `
<div
  className="SignIn-Content"
>
  <section
    className="Modal-Close"
  >
    <button
      className="closeBtn"
      onClick={[MockFunction]}
    >
      X
    </button>
  </section>
  <section
    className="Modal-Header"
  >
    <button
      className="btnChoice choice-inactive"
      onClick={[Function]}
    >
      Sign In
    </button>
    <button
      className="btnChoice choice-active"
      onClick={[Function]}
    >
      Create Account
    </button>
  </section>
  <section
    className="Modal-Body"
  >
    <form
      className="SignForm"
      onSubmit={[Function]}
    >
      <label>
        Email
      </label>
      <input
        className="email"
        onChange={[Function]}
        type="text"
      />
      <label>
        Password
      </label>
      <input
        className="password"
        onChange={[Function]}
        type="text"
      />
      <button
        type="submit"
      >
        Sign In
      </button>
    </form>
  </section>
</div>
`;
